typescript/sdk/src/token/checker.ts

```diff
@@ -86,37 +111,119 @@ export class HypERC20Checker extends ProxiedRouterChecker<
     } else if (isSyntheticConfig(expectedConfig)) {
       await checkERC20(hypToken as unknown as ERC20, expectedConfig);
     } else if (isCollateralConfig(expectedConfig)) {
+      const collateralToken = await this.getCollateralToken(chain);
+      const actualToken = await (
+        hypToken as unknown as HypERC20Collateral
+      ).wrappedToken();
+      if (!eqAddress(collateralToken.address, actualToken)) {
+        const violation: TokenMismatchViolation = {
+          type: 'CollateralTokenMismatch',
+          chain,
+          expected: collateralToken.address,
+          actual: actualToken,
+          tokenAddress: hypToken.address,
+        };
+        this.addViolation(violation);
+      }
+    }
+  }
+
+  private cachedAllActualDecimals: Record<ChainName, number> | undefined =
+    undefined;
+
+  async getEvmActualDecimals(): Promise<Record<ChainName, number>> {
+    if (this.cachedAllActualDecimals) {
+      return this.cachedAllActualDecimals;
+    }
+    const entries = await Promise.all(
+      this.getEvmChains().map(async (chain) => {
+        const token = this.app.router(this.app.getContracts(chain));
+        return [chain, await this.getActualDecimals(chain, token)];
+      }),
+    );
+
+    this.cachedAllActualDecimals = Object.fromEntries(entries);
+
+    return Object.fromEntries(entries);
+  }
+
+  async getActualDecimals(
+    chain: ChainName,
+    hypToken: TokenRouter,
+  ): Promise<number> {
+    const expectedConfig = this.configMap[chain];
+    let decimals: number | undefined = undefined;
+
+    if (isNativeConfig(expectedConfig)) {
+      decimals =
+        this.multiProvider.getChainMetadata(chain).nativeToken?.decimals;
+    } else if (isSyntheticConfig(expectedConfig)) {
+      decimals = await (hypToken as unknown as ERC20).decimals();
```

Comment:
Could this cast be avoided if `isSyntheticConfig` had a `: x is ERC20` type clause on it?

Reply:
think this would require the expectedConfig coming from hypToken or passing hypToken into `isSyntheticConfig` - for now I'll keep as is (this was what we were doing elsewhere before this PR too)